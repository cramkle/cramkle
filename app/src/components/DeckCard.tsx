import React from 'react'
import { useHistory } from 'react-router'
import Card, { CardPrimaryContent } from 'views/Card'
import { Body2, Headline6 } from 'views/Typography'

interface Props {
  title: string
  description?: string | null
  slug: string
}

const DeckCard: React.FunctionComponent<Props> = ({
  title,
  description = null,
  slug,
}) => {
  const history = useHistory()
  const handleClick = () => history.push(`/d/${slug}`)

  return (
    <Card outlined className="h-100">
      <CardPrimaryContent
        className="pa2 h-100"
        tabIndex={0}
        role="article"
        onClick={handleClick}
        onKeyDown={(e: React.KeyboardEvent) =>
          e.key === 'Enter' && handleClick()
        }
      >
        <Headline6>{title}</Headline6>
        {description && <Body2>{description}</Body2>}
      </CardPrimaryContent>
    </Card>
  )
}

export default DeckCard
