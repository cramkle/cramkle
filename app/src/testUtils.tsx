import { MockedProvider, MockedResponse } from '@apollo/react-testing'
import { setupI18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { render as rtlRender } from '@testing-library/react'
import { en as enPlural } from 'make-plural/plurals'
import React, { ReactElement } from 'react'
import { MemoryRouter } from 'react-router'

import enCatalog from './locales/en/messages'

interface RenderOptions {
  mocks?: MockedResponse[]
}

export function render(ui: ReactElement, options?: RenderOptions) {
  const i18n = setupI18n()

  i18n.load('en', enCatalog.messages)
  i18n.loadLocaleData({ en: { plurals: enPlural } })

  i18n.activate('en')

  const utils = rtlRender(
    <MemoryRouter>
      <I18nProvider i18n={i18n}>
        <MockedProvider mocks={options?.mocks} addTypename={false}>
          {ui}
        </MockedProvider>
      </I18nProvider>
    </MemoryRouter>
  )

  return utils
}
