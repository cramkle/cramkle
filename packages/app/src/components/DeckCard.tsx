import React from 'react'
import { withRouter, RouteComponentProps } from 'react-router'
import Card, { CardPrimaryContent } from '@material/react-card'
import { Body2, Headline6 } from '@material/react-typography'

interface Props extends RouteComponentProps {
  title: string
  description?: string
  slug: string
}

const DeckCard: React.FunctionComponent<Props> = ({
  title,
  description,
  slug,
  history,
}) => (
  <Card outlined className="h-100">
    <CardPrimaryContent
      className="pa2 h-100"
      tabIndex={0}
      role="button"
      onClick={() => history.push(`/d/${slug}`)}
    >
      <Headline6>{title}</Headline6>
      {description && <Body2>{description}</Body2>}
    </CardPrimaryContent>
  </Card>
)

DeckCard.defaultProps = {
  description: null,
}

export default withRouter(DeckCard)
