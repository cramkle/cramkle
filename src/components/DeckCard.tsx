import { plural } from '@lingui/macro'
import { useLingui } from '@lingui/react'
import gql from 'graphql-tag'
import React from 'react'
import { useHistory } from 'react-router'

import { FlashCardStatus } from '../globalTypes'
import FlashCardStatusChip from './FlashCardStatus'
import { DeckCard_deck } from './__generated__/DeckCard_deck'
import { Card, CardPressable } from './views/Card'
import { Body2, Headline3 } from './views/Typography'

interface Props {
  deck: DeckCard_deck
  onClick?: (deck: DeckCard_deck) => void
  showStudySessionDetails?: boolean
}

export const deckCardFragment = gql`
  fragment DeckCard_deck on Deck {
    id
    slug
    title
    description
    totalNotes
    totalFlashcards
    studySessionDetails {
      newCount
      learningCount
      reviewCount
    }
  }
`

const DeckCard: React.FunctionComponent<Props> = ({
  onClick,
  deck,
  showStudySessionDetails = false,
}) => {
  const history = useHistory()
  const { i18n } = useLingui()

  const handleClick = () => {
    if (onClick) {
      onClick(deck)
    } else {
      history.push(`/d/${deck.slug}`)
    }
  }

  const {
    studySessionDetails: { newCount, learningCount, reviewCount },
  } = deck

  const uniqueId = `deck-${deck.id}`

  return (
    <Card lean className="h-full">
      <CardPressable
        className="p-2 h-full"
        tabIndex={0}
        role="article"
        onClick={handleClick}
        onKeyDown={(e: React.KeyboardEvent) =>
          e.key === 'Enter' && handleClick()
        }
        aria-describedby={`${uniqueId}-title`}
      >
        <Headline3 id={`${uniqueId}-title`}>{deck.title}</Headline3>
        {deck.description && <Body2>{deck.description}</Body2>}
        {showStudySessionDetails ? (
          <div className="mt-1 flex">
            {newCount > 0 && (
              <FlashCardStatusChip
                status={FlashCardStatus.NEW}
                className="mr-2"
                truncated
              >
                {i18n._(plural(newCount, { one: '# new', other: '# new' }))}
              </FlashCardStatusChip>
            )}
            {learningCount > 0 && (
              <FlashCardStatusChip
                status={FlashCardStatus.LEARNING}
                className="mr-2"
                truncated
              >
                {i18n._(
                  plural(learningCount, {
                    one: '# learning',
                    other: '# learning',
                  })
                )}
              </FlashCardStatusChip>
            )}
            {reviewCount > 0 && (
              <FlashCardStatusChip status={FlashCardStatus.REVIEW} truncated>
                {i18n._(
                  plural(reviewCount, { one: '# review', other: '# review' })
                )}
              </FlashCardStatusChip>
            )}
          </div>
        ) : (
          <Body2 className="flex items-center">
            {i18n._(
              plural(deck.totalNotes, {
                one: '# note',
                other: '# notes',
              })
            )}
            <span className="inline-block mx-1">&middot;</span>
            {i18n._(
              plural(deck.totalFlashcards, {
                one: '# flashcard',
                other: '# flashcards',
              })
            )}
          </Body2>
        )}
      </CardPressable>
    </Card>
  )
}

export default DeckCard
