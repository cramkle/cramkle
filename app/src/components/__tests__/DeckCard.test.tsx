import React from 'react'

import { render } from '../../testUtils'
import DeckCard from '../DeckCard'

describe('<DeckCard />', () => {
  it('should match snapshot', () => {
    const comp = render(
      <DeckCard
        deck={{
          __typename: 'Deck',
          id: '123',
          title: '日本語げんき',
          description: '',
          slug: 'nihongo-genki',
          studySessionDetails: {
            __typename: 'StudySessionDetails',
            newCount: 0,
            learningCount: 0,
            reviewCount: 0,
          },
        }}
      />
    )

    expect(comp.container).toMatchSnapshot()
  })
})
