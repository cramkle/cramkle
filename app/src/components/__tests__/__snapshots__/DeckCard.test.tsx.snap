// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeckCard /> should match snapshot 1`] = `
<div>
  <div
    class="mdc-card h-full mdc-card--outlined"
  >
    <div
      class="mdc-card__primary-action p-2 h-full"
      role="article"
      tabindex="0"
    >
      <h6
        class="mdc-typography mdc-typography--headline6"
      >
        日本語げんき
      </h6>
    </div>
  </div>
</div>
`;
